File: message-queue/rust/src/lib/prometheus.rs
use std::error;

use prometheus::{IntGauge, Registry, Gauge};

use crate::Instance;
use crate::runtime::Runtime;
use crate::SampleDesc;

pub struct Prometheus {
    registry: Registry,
    pub test_metric: IntGauge,
    pub messages_total: IntGauge,
    pub messages_per_second: Gauge,
    pub duration_seconds: Gauge,
}

impl Prometheus {
    pub fn new(rt: &Runtime) -> Result<Self,Box<dyn error::Error>> {
        let r = Registry::new_custom(None, Some(rt.to_map()));
        if let Err(e) = r {
            return Err(Box::new(e));
        }

        let s = Self{
            registry: r.unwrap(),
            test_metric: IntGauge::new("test", "Test metric.").unwrap(),
            messages_total: IntGauge::new(
                "messages_total",
                "Messages sent."
            ).unwrap(),
            messages_per_second: Gauge::new(
                "messages_per_second",
                "Messages per second sent."
            ).unwrap(),
            duration_seconds: Gauge::new(
                "duration_seconds",
                "Work duration in seconds."
            ).unwrap(),
        };
        s.registry.register(Box::new(s.test_metric.clone()))?;
        s.registry.register(Box::new(s.messages_total.clone()))?;
        s.registry.register(Box::new(s.messages_per_second.clone()))?;
        s.registry.register(Box::new(s.duration_seconds.clone()))?;
        return Ok(s);
    }

    pub fn push(&self, sdesc: &SampleDesc, worker_id: u64) -> Result<(),Box<dyn error::Error>> {
        let metrics = self.registry.gather();
        let mut labels = sdesc.to_map();
        labels.insert("worker_id".to_string(), worker_id.to_string());
        let r = prometheus::push_add_metrics(
            "mq-producer",
            labels,
            "localhost:9091",
            metrics,
            None,
        );
        if let Err(e) = r {
            return Err(Box::new(e));
        }

        return Ok(());
    }
}

pub fn test_cmd(app: &Instance) -> Result<(), Box<dyn error::Error>> {
    println!("Testing prometheus");
    println!("{:?}", app.runtime.to_map());

    let sdesc = SampleDesc{
        n_workers: 4,
        algorithm: "threading",
        mq_system: "postgres",
    };

    app.prometheus.test_metric.set(2);
    app.prometheus.push(&sdesc, 1)?;

    return Ok(());
}

File: message-queue/golang/src/prometheus.go
package jmcgmqp

import "time"
import "strconv"
import "fmt"
import "github.com/prometheus/client_golang/prometheus"
import "github.com/prometheus/client_golang/prometheus/push"
import "golang.org/x/exp/maps"

var (
	TestMetric = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "test",
		Help: "Test metric",
	},
		append(append([]string{"worker_id"}, RuntimeFieldNames()...), SampleDescFieldNames()...),
	)

	MessagesTotal = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "messages_total",
		Help: "Messages sent",
	},
		append(append([]string{"worker_id"}, RuntimeFieldNames()...), SampleDescFieldNames()...),
	)

	MessagesPerSecond = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "messages_per_second",
		Help: "Messages per second sent",
	},
		append(append([]string{"worker_id"}, RuntimeFieldNames()...), SampleDescFieldNames()...),
	)

	DurationSeconds = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "duration_seconds",
		Help: "Work duration in seconds",
	},
		append(append([]string{"worker_id"}, RuntimeFieldNames()...), SampleDescFieldNames()...),
	)
)

func NewPusher() *push.Pusher {
	registry := prometheus.NewRegistry()
	registry.MustRegister(TestMetric)
	registry.MustRegister(MessagesTotal)
	registry.MustRegister(MessagesPerSecond)
	registry.MustRegister(DurationSeconds)

	pusher := push.New("http://localhost:9091", "mq-producer").Gatherer(registry)
	return pusher
}

func mkLabels(app *Instance, sample SampleDesc, worker *WorkerResult) prometheus.Labels {
	str_labels := app.Runtime.Map()
	maps.Copy(str_labels, map[string]string{"worker_id": strconv.Itoa(worker.WorkerId)})
	maps.Copy(str_labels, sample.Map())
	var labels prometheus.Labels = str_labels
	return labels
}

func TestPusher(app *Instance) {
	fmt.Println("Testing prometheus push")
	result := NewWorkerResult(0, 1, time.Second)
	labels := mkLabels(app, SampleDesc{8, "goroutines", "postgres"}, result)
	TestMetric.With(labels).Set(42)
	if err := app.Prometheus.Add(); err != nil {
		panic(fmt.Sprintf("Prometheus push failed: %v", err.Error()))
	}
}

func PushMetrics(app *Instance, sample SampleDesc, rs *Results) {
	for _, r := range rs.Workers {
		MessagesTotal.With(mkLabels(app, sample, r)).Set(float64(r.MessagesTotal))
		MessagesPerSecond.With(mkLabels(app, sample, r)).Set(r.MessagesPerSecond)
		DurationSeconds.With(mkLabels(app, sample, r)).Set(r.DurationSeconds)

		if err := app.Prometheus.Add(); err != nil {
			panic(fmt.Sprintf("Prometheus push failed: %v", err.Error()))
		}
	}
}
